src/parser_urdf.cc

```diff
@@ -2663,48 +2665,133 @@ void URDF2SDF::ListSDFExtensions(const std::string &_reference)
+    {
+      errorStream << "no <inertial> block defined. ";
+    }
+    else
     {
-      sdfdbg << "urdf2sdf: link[" << _link->name
-             << "] has no inertia, ["
-             << static_cast<int>(_link->child_links.size())
-             << "] children links ignored.\n";
+      errorStream << "a mass value of less than or equal to zero. ";
     }
+    errorStream << "Please ensure this link has a valid mass to prevent any "
+                << "missing links or joints in the resulting SDF model.";
+    nonJointReductionErrors.emplace_back(
+        ErrorCode::LINK_INERTIA_INVALID, errorStream.str());
+    errorStream.str(std::string());
+
+    // If joint reduction happens and resolves the issue of this link with no
+    // mass, no warnings will be emitted
+    bool jointReductionHappens = _link->parent_joint &&
+        FixedJointShouldBeReduced(_link->parent_joint) &&
+        g_reduceFixedJoints;
 
-    if (!_link->child_joints.empty())
+    // check if parent joint can be lumped and reduced
+    if (!jointReductionHappens && _link->parent_joint)
     {
-      sdfdbg << "urdf2sdf: link[" << _link->name
-             << "] has no inertia, ["
-             << static_cast<int>(_link->child_links.size())
-             << "] children joints ignored.\n";
+      if (_link->parent_joint->type == urdf::Joint::FIXED)
+      {
+        errorStream << "urdf2sdf: allowing joint lumping by removing any "
+                    << "<disableFixedJointLumping> or <preserveFixedJoint> "
+                    << "gazebo tag on fixed parent joint["
+                    << _link->parent_joint->name
+                    << "], as well as ensuring that "
+                    << "ParserConfig::URDFPreserveFixedJoint is false, could "
+                    << "help resolve this warning. See http://sdformat.org/tuto"
+                    << "rials?tut=sdformat_urdf_extensions&cat=specification& "
+                    << "for more information about this behavior.";
+        nonJointReductionErrors.emplace_back(
+            ErrorCode::LINK_INERTIA_INVALID, errorStream.str());
+        errorStream.str(std::string());
+      }
+
+      errorStream << "urdf2sdf: parent joint["
+                  << _link->parent_joint->name
+                  << "] ignored.";
+      nonJointReductionErrors.emplace_back(
+          ErrorCode::LINK_INERTIA_INVALID, errorStream.str());
+      errorStream.str(std::string());
     }
 
-    if (_link->parent_joint)
+    // check child joints for any possibility of joint lumping and reduction
+    if (!jointReductionHappens && !_link->child_joints.empty())
     {
-      sdfdbg << "urdf2sdf: link[" << _link->name
-             << "] has no inertia, "
-             << "parent joint [" << _link->parent_joint->name
-             << "] ignored.\n";
+      for (const auto &cj : _link->child_joints)
+      {
+        if (FixedJointShouldBeReduced(cj) && g_reduceFixedJoints)
+        {
+          jointReductionHappens = true;
+          break;
```

Comment:
I believe so! This test will cause the `break`, https://github.com/gazebosim/sdformat/pull/1238/files#diff-fc2484863e209b4368da90c08381488ec4b010f386c642b1ea1e8bcf82cf6f02R2141-R2190

Could this be due to the code only checking `!jointReductionHappens` and not do anything if `jointReductionHappens` is true?

I'm not too sure how to see if this gets reported again, is it just https://app.codecov.io/gh/gazebosim/sdformat/pull/1238? Looks alright at the moment

Reply:
I tested with gdb, and looks like it really isn't covered. Apparently, a massless link with a non-fixed parent joint and a fixed child joint is reduced before `CreateSDF` is ever called : https://github.com/gazebosim/sdformat/blob/ba52e6fe48e0c236a5ccf07f4b33164662832a43/src/parser_urdf.cc#L3385-L3398

This means, the massless link now has mass from the lumped child links. So, I guess if we encounter a massless link here, it would only be if:
  * it's parent joint is reduced (i.e, the link has been lumped into it's parent link) or
  * it's child joint is fixed, but could not be reduced because joint lumping is disabled.

Can you verify my logic? Does that maybe simplify anything?

Reply:
Yup, I can recreate it with `gdb` too, it doesn't even allow me to set a breakpoint there, probably got optimized away by the compiler. Looking at the code, I can verify that is the intended behavior too.

https://github.com/gazebosim/sdformat/pull/1238/commits/94382d7a1c692076ca20610d623434b0db7d0a4d, https://github.com/gazebosim/sdformat/pull/1238/commits/6aa625c83e3326f974835f2c6ca7ccd71d59078c, I've refactored the fixed child joint logic and reduced the number of checks we are doing overall (only if parent joint is reduced)

I kept the use of `Errors`, instead of using `sdfwarn << Error(...)` directly, to make easier when we do introduce passing `Errors` into these functions in the future.

Reply:
@azeey, per VC, regarding whether or not a similar situation for reduced parent joints is a bug, I don't believe it is a bug, just that checks happen after . Before this PR, joints are lumped recursively onto their parents with L3387 (shared by you above), and relies on https://github.com/gazebosim/sdformat/pull/1238/files#diff-2bc5ca23bcfc66fe173f513399bec8065b1cb4e607d741566517d7368e6bce0fL2705 to not create the link (whose's inertia has been lumped) if the parent joint is reduced, but the urdf link still contains the mass values

so there will be cases where `CreateSDF` is called on a massless link, https://github.com/gazebosim/sdformat/pull/1238/files#diff-2bc5ca23bcfc66fe173f513399bec8065b1cb4e607d741566517d7368e6bce0fL2713, where the parent has already been reduced, hence the need to check if the parent has been reduced.

on the otherhand, as you surmised, a massless link `A` with a fixed child joint, and child link `B` with mass, will already have `B`'s mass lumped into it in `CreateSDF`, so it should always have mass unless joint lumping was turned off.